### ISSUE TITLE
about how to get raw camera data by colmap

### ISSUE THREAD

**Issue by tscvividly** [adobe-research/VideoDoodles#6]
I use colmap to get rotation and translation, but these data are different from your data. Could you please tell me in detail how to use colmap to obtain detailed pose information?

**Comment by SongOnWater** [comment]
also curious about this question ,how to the get the data of refined_cameras.txt

**Comment by em-yu** [comment]
Hi, we used Adobe in-house depth and camera optimisation code to process Colmap and video data. Since we were not authorized to release that code as part of this work, there is indeed a missing link between colmap outputs and our inputs.

Even I do not personally have access to that code anymore, but looking at the colmap output format described [here](https://colmap.github.io/format.html#images-txt) and looking at our input format described in the readme and [here](https://github.com/adobe-research/VideoDoodles/blob/main/preprocess/videodepth_camera.py#L18) in the class, I think the following mapping is what is happening in terms of conversion:

With this being the colmap output format:

```
# In cameras.txt:
# Camera list with one line of data per camera:
#   CAMERA_ID, MODEL, WIDTH, HEIGHT, PARAMS[]
1 PINHOLE 3072 2304 2560.56 2560.56 1536 1152

# In images.txt:
# Image list with two lines of data per image:
#   IMAGE_ID, QW, QX, QY, QZ, TX, TY, TZ, CAMERA_ID, NAME
#   POINTS2D[] as (X, Y, POINT3D_ID)
```

And this being our format:
```
t_x t_y t_z rot_x rot_y rot_z scale shift
# one line like this per image
# ...

# At the end of the file:
fx fy
```

`t_x` etc correspond to `TX` for each image, and `rot_x` etc correspond to the equivalent of the `QW, QX, QY, QZ` quaternion values but converted to a [rotation vector](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.from_rotvec.html).

`fx` and `fy` correspond to the parameters in `cameras.txt` (here `Fx = 2560.56` and  `Fy = 2560.56`) that are converted to normalized device coordinates (`fx = 2 * Fx / width`). This is something I'm quickly re-understanding based off of my [conversion code](https://github.com/adobe-research/VideoDoodles/blob/main/preprocess/videodepth_camera.py#L63), I have not thoroughly thought about this or tested it, so please take that with a grain of salt.

Also, compared to raw colmap data, the (unreleased) processing stage also optimized both depth and camera values to make them smoother.

Sorry that only partially answers that question. Good luck!

**Comment by SongOnWater** [comment]
Thanks for your responses. I just wonder that the fx,fy values in refined_cameras.txt  of the raw samples seem  like less than 10.0 , such as:  "5.005727616480021	8.912274688094993"  or  "2.8197032239516457	5.020243128697292" , but the values got from Colmap always more than 1000 , such as "Fx = 2560.56 and Fy = 2560.56 " , how to convert the Colmap big value to the raw data small value, if the code is not accessible , could you show the idea about the implementation.

**Comment by em-yu** [comment]
The fx, fy values are normalized by the resolution of the image. So to transform those from Colmap you have to do: `fx = 2 * Fx / width` with `width` the number of pixels in the width of the image.

The idea of the implementation is based on this paper: [Consistent Video Depth Estimation](https://roxanneluo.github.io/Consistent-Video-Depth-Estimation/).

**Comment by SongOnWater** [comment]
Thanks for your guidance. There is another confusion for me. For "resized_disps.npz" the processing doc says it "inferred with a re-implementation of [Robust Consistent Video Depth Estimation](https://robust-cvd.github.io/).", so does the "re-implementation" also means the code not accessible? As I tried to run Robust_cvd for  getting the "disparity map", I got the result not as same as  "resized_disps.npz" in raw samples. And I noticed that resized_disps.npz  seems just converted to depth map for further processing by  multiplicative inversing. So I doubt it got by  multiplicative inverse depth map. Could you offer more detail about how to get the data of  "resized_disps.npz" ?

**Comment by em-yu** [comment]
The code for this re-implementation is not available unfortunately. I personally don't have access to it anymore, and I was not involved in the implementation (this was work done by Kevin Blackburn-Matzen at Adobe). The depth maps were definitely better for our purpose than the ones we obtained from RCVD (which we tried too). `resized_disps` contains disparity maps, that's why there is an inverse operation to obtain the depth map

**Comment by SongOnWater** [comment]
Thanks for your responses. Thanks for your response. I have also tried to run Colmap with the frames obtained from raw data examples, but the data I obtained seems to have large differences with the data of the raw data example in refined_cameras.txt. So, are there any specific parameters for Colmap? As I am not familiar with Colmap, it would be very appreciated if I could get more detail about the process.

**Comment by SongOnWater** [comment]
A small question that if fx = 2 * Fx / width, so fy =2 * Fy / height , right?

**Comment by ouououououyy** [comment]
Hi, Did you succeed in making it?

**Comment by SongOnWater** [comment]
No. I can't get a qualifiable  refined_cameras.txt to  make the drawing move with the target . But recently may advanced video generation model have improved, such as Runway , Wan VACE, or even Blender or AE have AI plugin may help  , or maybe you can implement your creative idea with other tools.

**Comment by ouououououyy** [comment]
I have used these video tools before. However, it seems that the controllability and stability of vediodoodles are better. So are you still trying to complete the work of converting vedio to raw data?

**Comment by SongOnWater** [comment]
No. I had tried to reimplement the processes, but failed. It had costed lot of my time, maybe as I am not knowledgeable in these area. I just gave up.

**Comment by em-yu** [comment]
Hi, sorry for the lack of replies, I don't have much concrete help to offer. I thought I'd share some more thoughts in case that's useful.

The complete pipeline for cameras+depth estimation and conversion was definitely a lot of work for us to put together, and I also find it unfortunate that we cannot share end-to-end (this is an IP-related compromise). I think one could build back something similar, but it will take some trial-and-error in order to resolve potential mismatches in camera models, and depth map format. I tried to use common conventions, such as the COLMAP camera model, but there could be subtleties that I haven't taken into account since I always used the same source of raw data. 

In my experience, what was helpful in the process of making sense of camera+depth data was breaking down the testing into small steps (eg: first make sure that the camera extrinsics is plausible by visualising the [3D positions](https://github.com/adobe-research/VideoDoodles/blob/main/preprocess/videodepth_camera.py#L106) and the [orientations](https://github.com/adobe-research/VideoDoodles/blob/main/preprocess/videodepth_camera.py#L109) in space, then visualize the depth maps images in 2D, finally [compute un-projected pixel positions](https://github.com/adobe-research/VideoDoodles/blob/main/preprocess/videodepth_camera.py#L93) in 3D based on camera+depth and visually inspect that point cloud). For a start, I would definitely suggest ditching the frontend and using purely the Python scripts to debug. In Python it's possible to rely on a simple 3D library to get debug visualizations (eg Polyscope). This will help diagnose where the conversion mismatches come from. All that said, I can't say for sure doing all this work is the right call for everyone, and I think it's wise to consider dropping the matter.

**Comment by SongOnWater** [comment]
https://github.com/yukangcao/Awesome-4D-Spatial-Intelligence , hope this page could help you
